stages/org.osbuild.copy

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python3
+
+import json
+import os
+import sys
+import subprocess
+
+import osbuild.sources
+
+
+STAGE_DESC = "Copy files from a source to the tree"
+STAGE_INFO = """
+Copies files obtained via a `source` to the tree. Multiple files or
+directories can be copied by specifying multiple entries in `paths`.
+If no paths are specified the whole contents of `source` is copied.
+The source and the target path for each individual path entry might
+optionally be specified via `from` and `to`, respectively; of no
+path is given for any of the two, the root `/` is assumed.
+
+Supported sources are currently:
+
+ o `git`: Fetch the contents of a specific git commit. The commit
+   is specified via its sha1 in the `commit` option. Internally
+   uses the 'org.osbuild.git' source to fetch the commit.
+
+"""
+STAGE_OPTS = """
+"additionalProperties": false,
+"definitions": {
+  "source-git": {
+    "description": "Fetch a checkout via git",
+    "type": "object",
+    "required": ["type", "commit"],
+    "additionalProperties": false,
+    "properties": {
+      "type": {
+        "type": "string",
+         "description": "The type of the source, here 'git'",
+         "enum": ["git"]
+      },
+      "commit": {
+        "type": "string",
+         "description": "The specific commit to fetch"
+      }
+    }
+  }
+},
+"properties": {
+  "source": {
+    "description": "Method of how to obtain the source data",
+    "oneOf": [{ "$ref": "#/definitions/source-git" }]
+  },
+  "paths": {
+    "description": "Array of items to copy",
+    "type": "array",
+    "items": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "from": {
+          "type": "string",
+           "description": "The source path",
+           "default": "/"
+        },
+        "to": {
+          "type": "string",
+           "description": "The destination path",
+           "default": "/"
+        }
+      }
+    }
+  }
+}
+"""
+
+
+def get_git(options, srcdir):
+    commit = options["commit"]
+    res = osbuild.sources.get("org.osbuild.git", [commit])
+    source_path = os.path.join(srcdir, "org.osbuild.git")
+    return source_path, res
+
+
+def main(tree, srcdir, options):
+    source_options = options["source"]
+    source_type = source_options["type"]
+    paths = options.get("paths", [{}])
+
+    if source_type == "git":
+        source = get_git
+    else:
+        raise ValueError(f"Unsupported source: {source_type}")
+
+    # Fetch the requested data via the osbuild source
+    source_path, _ = source(source_options, srcdir)
+
+    for path in paths:
+        src = path.get("from", "/")
+        dst = path.get("to", "/")
+
+        subprocess.run(["cp", "--reflink=auto", "-a",
+                        f"{source_path}{src}",
+                        f"{tree}{dst}"],
+                       check=True)
```

Comment:
I like this very much. Quite flexible and we can add more types in the future.

Can you remind me why we never used `shutil.copytree()`? Is this to preserve metadata better?

Reply:
I think it did not handle reflinks properly, but @teg might remember better.

Reply:
Yeah, it was about preserving metadata. I forgot the details though.

Reply:
From the `shutil` [docs](https://docs.python.org/3/library/shutil.html):
```
Warning

Even the higher-level file copying functions (shutil.copy(), shutil.copy2()) cannot copy all file metadata.
```